sycl/doc/Assert.md

```diff
@@ -0,0 +1,250 @@
+# Assert feature
+
+**IMPORTANT**: This document is a draft.
+
+During debugging of kernel code user may put assertions here and there.
+The expected behaviour of assertion failure at host is application abort.
+Our choice for device-side assertions is asynchronous exception in order to
+allow for extensibility.
+
+The user is free to disable assertions by defining `NDEBUG` macro at
+compile-time.
+
+
+## Use-case example
+
+```
+using namespace cl::sycl;
+auto ErrorHandler = [] (exception_list Exs) {
+  for (exception_ptr const& E : Exs) {
+    try {
+      std::rethrow_exception(E);
+    }
+    catch (event_error const& Ex) {
+      std::cout << “Exception - ” << Ex.what(); // assertion failed
+      std::abort();
+    }
+  }
+};
+
+void user_func(item<2> Item) {
+  assert((Item[0] % 2) && “Nil”);
+}
+
+int main() {
+  queue Q(ErrorHandler);
+  q.submit([&] (handler& CGH) {
+    CGH.parallel_for<class TheKernel>(range<2>{N, M}, [=](item<2> It) {
+      do_smth();
+      user_func(It);
+      do_smth_else();
+    });
+  });
+  Q.wait_and_throw();
+  std::cout << “One shouldn’t see this message.“;
+  return 0;
+}
+```
+
+In this use-case every work-item with even X dimension will trigger assertion
+failure. Assertion failure should be reported via asynchronous exceptions. If
+asynchronous exception handler is set the failure is reported with
+`cl::sycl::event_error` exception. Otherwise, SYCL Runtime should trigger abort.
```

Comment:
Consider adding a new exception class specific for assert, it should be distinguishable from other exceptions.

Reply:
See my comment elsewhere.  I suggest basing this extension spec on the SYCL 2020 exception model, where all SYCL exceptions have the same class.  I suggest creating a new exception error code to distinguish assertion failures from other exceptions.

Reply:
Yes, I think an assertion failure should throw some new exception code.  I think @kbobrovs is saying that also.

Reply:
Added extension document with exception error code.